python/katana/local/analytics/_cdlp.pyx

```diff
@@ -0,0 +1,247 @@
+
+    Result[void] CdlpAssertValid(_PropertyGraph*pg, string output_property_name)
+
+    cppclass _CdlpStatistics "katana::analytics::CdlpStatistics":
+        uint64_t total_communities
+        uint64_t total_non_trivial_communities
+        uint64_t largest_community_size
+        double largest_community_ratio
+
+        void Print(ostream os)
+
+        @staticmethod
+        Result[_CdlpStatistics] Compute(_PropertyGraph*pg, string output_property_name)
+
+
+class _CdlpPlanAlgorithm(Enum):
+    """
+    :see: :py:class:`~katana.local.analytics.CdlpPlan` constructors for algorithm documentation.
+    """
+    Synchronous = _CdlpPlan.Algorithm.kSynchronous
+    #Asynchronous = _CdlpPlan.Algorithm.kAsynchronous
+
+
+cdef class CdlpPlan(Plan):
+    """
+    A computational :ref:`Plan` for CDLP.
+
+    Static method construct CdlpPlans using specific algorithms with their required parameters. All
+    parameters are optional and have reasonable defaults.
+    """
+    cdef:
+        _CdlpPlan underlying_
+
+    cdef _Plan* underlying(self) except NULL:
+        return &self.underlying_
+
+    Algorithm = _CdlpPlanAlgorithm
+
+    @staticmethod
+    cdef CdlpPlan make(_CdlpPlan u):
+        f = <CdlpPlan> CdlpPlan.__new__(CdlpPlan)
+        f.underlying_ = u
+        return f
+
+    @property
+    def algorithm(self) -> _CdlpPlanAlgorithm:
+        return self.underlying_.algorithm()
+
+    @staticmethod
+    def synchronous() -> CdlpPlan:
+        """
+        Initially, all nodes are in their own community IDs (same as their
+        node IDs). Then, the community IDs are iteratively set to the most
+        frequent community ID in their immediate neighborhood. It continues
+        untill the community ID of all nodes in graph become the same as
+        the most frequent ID in their immediate neighborhood.
+
+        Synchronous community detection algorithm. This algorithm is based on
+        Graphalytics benchmark that has two key differences from the original algorithm
+        proposed in [1]. First, it is deterministic: if there are multiple
+        labels with their frequency equalling the maximum, it selects the smallest
+        one while the original algorithm selects randomly. Second, it is synchronous,
+        i.e., each iteration is computed based on the labels obtained as a result of
+        the previous iteration.
+
+        FIXME: As remarked in [1], this can cause the oscillation
+        of labels in bipartite or nearly bipartite subgraphs. This is especially true
+        in cases where communities take the form of a star graph
+        """
+        return CdlpPlan.make(_CdlpPlan.Synchronous())
+
+    #@staticmethod
+    #def asynchronous() -> CdlpPlan:
+        """
+        Unlike Synchronous algorithm, Asynchronous can use the current iteration
+        updated community IDs for some of the neighbors that have been already
+        updated in the current iteration and use the old values for the other neighbors
```

Comment:
If you are commenting code, comment the whole thing. ;-)

Reply:
You are right; I have been more a c++ guy; I considered what is inside the triple quote as comments. will comment the whole thing; thanks for the comment.

Reply:
Yep. :-) That's what I wanted

Reply:
I think also about maybe removing these redundant comments here; these comments already are in the CPP implementation of each algorithm.